src/Revisionable/Mapping/Driver/Attribute.php

```diff
@@ -0,0 +1,172 @@
+                }
+
+                if (!$cl = $this->getRelatedClassName($meta, $annot->revisionClass)) {
+                    throw new InvalidMappingException(sprintf("The revision class '%s' configured for '%s' does not exist.", $class, $meta->getName()));
+                }
+
+                $config['revisionClass'] = $cl;
+            }
+        }
+
+        // Inspect properties for versioned fields
+        foreach ($class->getProperties() as $property) {
+            $field = $property->getName();
+
+            if ($this->reader->getPropertyAnnotation($property, Versioned::class)) {
+                if ($meta->isCollectionValuedAssociation($field)) {
+                    throw new InvalidMappingException(sprintf('Cannot version field %s::$%s, collection valued associations are not supported.', $meta->getName(), $field));
+                }
+
+                // The MongoDB ODM's @EmbedMany is not supported
+                if ($meta instanceof MongoDBDOMClassMetadata && $meta->isCollectionValuedEmbed($field)) {
+                    throw new InvalidMappingException(sprintf('Cannot version field %s::$%s, an embedded many collection is not supported.', $meta->getName(), $field));
+                }
+
+                // To version a field with a relationship, it must be the owning side
+                if ($this->isRelationship($meta, $field)) {
+                    $associationMapping = $meta->associationMappings[$field];
+
+                    if (!$associationMapping['isOwningSide']) {
+                        throw new InvalidMappingException(sprintf('Cannot version field %s::$%s, it is not the owning side of the relationship.', $meta->getName(), $field));
+                    }
+                }
+
+                /*
+                 * Due to differences in the UoW's for each object manager, embedded models need to be handled differently.
+                 *
+                 * The MongoDB ODM tracks embedded documents within the UoW and the listener can recursively inspect the change set
+                 * for changes to these objects.
+                 *
+                 * The ORM inlines embedded field mappings to the root entity, so the list of versioned fields needs to be added to
+                 * the extension metadata now.
+                 */
+
+                if ($meta instanceof ORMClassMetadata && isset($meta->embeddedClasses[$field])) {
+                    $config = $this->inspectEmbeddedForVersioned($field, $config, $meta);
+
+                    continue;
+                }
+
+                $config['versioned'][] = $field;
+            }
+        }
+
+        // Validate configuration
+        if (!$meta->isMappedSuperclass && $config) {
+            // The revisionable flag must be set, except for embedded models, and the versioned config should be a non-empty array
+            if (isset($config['versioned']) && (!$this->isEmbed($meta) && !isset($config['revisionable']))) {
```

Comment:
Move the cheaper check first and remove the superfluos parentheses.

```suggestion
            if (isset($config['versioned']) && !isset($config['revisionable']) && !$this->isEmbed($meta)) {
```